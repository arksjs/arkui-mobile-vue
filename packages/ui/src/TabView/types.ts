import type { SwiperOnAnimated } from '../Swiper/types'

type OnChange = (name: string, index: number) => void

export interface TabViewProps {
  modelValue?: string
  initialVertical?: boolean
  scrollThreshold?: number
}

export interface TabViewEmits {
  onChange?: OnChange
  onAnimated?: SwiperOnAnimated
}

export interface TabViewItemProps {
  name: string
  title?: string
  subTitle?: string
}

export interface TabViewRef {
  switchTo: (name: string) => void
  switchToIndex: (index: number) => void
}

export type {
  OnChange as TabViewOnChange,
  SwiperOnAnimated as TabViewOnAnimated
}
