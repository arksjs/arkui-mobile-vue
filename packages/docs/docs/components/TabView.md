# TabView/TabViewItem 标签页

<CodeDemo name="TabView">

<<< @/../../demo/src/components/Navigation/TabView/index.vue

</CodeDemo>

## Import

```js
import { TabView, TabViewItem } from 'vfox'
```

具体的引入方式可以参考[引入组件](../guide/import.md)。

### Import Type

组件导出的类型定义：

```ts
import type { TabViewOnChange, TabViewOnAnimated } from 'vfox'
```

## TabView Props

| 属性                   | 类型    | 默认值 | 必填 | 说明                                            |
| ---------------------- | ------- | ------ | ---- | ----------------------------------------------- |
| initial-vertical       | boolean | false  | 否   | 初始化是否侧边菜单展示方式                      |
| scroll-threshold       | number  | 4      | 否   | 超过 `scrollThreshold` 个 Tab 使用滚动形式      |
| back-upper-when-change | boolean | false  | 否   | 切换面板时，如果是旧面板，是否返回顶部/左侧位置 |

## TabView Events

| 事件     | 描述           | 回调函数参数            | TypeScript 函数   |
| -------- | -------------- | ----------------------- | ----------------- |
| change   | 切换时触发     | ( activeIndex: number ) | TabViewOnChange   |
| animated | 动画结束时触发 | ( activeIndex: number ) | TabViewOnAnimated |

## TabView Slots

### #default

注：其中只可放置 [TabViewItem](./TabView.md#tabviewitem-props) 组件，否则会导致未定义的行为。

```vue
<fx-tab-view>
  <fx-tab-view-item name="Tab 1">
    <fx-empty description="Tab 1"/>
  </fx-tab-view-item>
  <fx-tab-view-item name="Tab 2">
    <fx-empty description="Tab 1" />
  </fx-tab-view-item>
</fx-tab-view>
```

## TabViewItem Props

| 属性 | 类型   | 默认值 | 必填 | 说明           |
| ---- | ------ | ------ | ---- | -------------- |
| name | string |        | 是   | 对应的菜单名称 |

## TabViewItem Slots

### 内容（#default）

```vue
<fx-tab-view-item name="Tab 1">
  <fx-empty description="Tab 1"/>
</fx-tab-view-item>
```
