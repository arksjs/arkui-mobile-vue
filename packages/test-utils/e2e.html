<script src="../../node_modules/vue/dist/vue.global.js"></script>
<script src="../../dist/index.js"></script>
<link rel="stylesheet" href="../../dist/index.css" />

<div id="app"></div>
