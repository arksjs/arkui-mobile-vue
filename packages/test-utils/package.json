{
  "name": "@arksjs/test-utils",
  "version": "0.0.1",
  "description": "arkui test-utils",
  "main": "dist/index.js",
  "private": true
}
