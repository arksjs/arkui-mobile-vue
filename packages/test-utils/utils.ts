export const timeout = (n: number) => new Promise(r => setTimeout(r, n))
