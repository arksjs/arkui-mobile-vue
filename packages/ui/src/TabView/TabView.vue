<template>
  <div :class="classes">
    <div class="fx-tab-view_header fx-horizontal-hairline">
      <SideTab
        v-if="vertical"
        :options="tabList"
        v-model:activeValue="activeIndex"
      />
      <Tab
        v-else
        :options="tabList"
        v-model:activeValue="activeIndex"
        :scroll-threshold="scrollThreshold"
      />
    </div>
    <div class="fx-tab-view_body" ref="listEl">
      <Swiper
        v-model:activeIndex="activeIndex"
        @change="onChange"
        @animated="onAnimated"
        ref="swiper"
        :initialVertical="vertical"
        :bounces="false"
      >
        <slot></slot>
      </Swiper>
    </div>
  </div>
</template>

<script lang="ts">
import { ref, defineComponent, provide } from 'vue'
import { Tab } from '../Tab'
import { SideTab } from '../SideTab'
import { Swiper } from '../Swiper'
import { useList } from '../hooks/use-list'
import type {
  OnChange as SwiperOnChange,
  OnAnimated as SwiperOnAnimated,
  SwiperRef
} from '../Swiper/types'
import { emitChangeValidator } from '../Swiper/props'
import { getClasses } from './util'
import type { PropsToEmits } from '../helpers/types'
import type { TabViewEmits } from './types'

export default defineComponent({
  name: 'fx-tab-view',
  components: { Tab, SideTab, Swiper },
  props: {
    // 纵向
    initialVertical: {
      type: Boolean,
      default: false
    },
    scrollThreshold: {
      type: Number,
      default: 4
    }
  },
  emits: {
    change: emitChangeValidator,
    animated: emitChangeValidator
  } as PropsToEmits<TabViewEmits>,
  setup(props, { emit }) {
    const vertical = ref<boolean>(!!props.initialVertical)
    const swiper = ref<SwiperRef>()
    const tabList = ref<
      {
        value: number
        label: string
        subLabel: string
      }[]
    >([])
    const activeIndex = ref(0)

    function resetItems($items: HTMLElement[]) {
      tabList.value = $items.map(($item, index) => {
        const { name, subName } = $item.dataset

        return {
          value: index,
          label: name || '',
          subLabel: subName || ''
        }
      })
    }

    const { listEl } = useList('tabView', resetItems)

    const onChange: SwiperOnChange = (activeIndex, fromIndex) => {
      emit('change', activeIndex, fromIndex)
    }

    const onAnimated: SwiperOnAnimated = (activeIndex, fromIndex) => {
      emit('animated', activeIndex, fromIndex)
    }

    function swipeTo(activeIndex: number) {
      swiper.value?.swipeTo(activeIndex)
    }

    provide('fxTabViewVertical', vertical.value)

    const classes = getClasses(vertical.value)

    return {
      activeIndex,
      tabList,
      vertical,
      listEl,
      swiper,
      onChange,
      onAnimated,
      swipeTo,
      classes
    }
  }
})
</script>
