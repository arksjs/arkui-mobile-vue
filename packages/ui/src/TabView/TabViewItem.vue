<template>
  <div
    class="fx-swiper-item fx-tab-view-item"
    :data-name="name"
    :data-sub-name="subName"
    ref="root"
  >
    <slot></slot>
  </div>
</template>

<script lang="ts">
import { ref, defineComponent, inject } from 'vue'
import { useListItem } from '../hooks/use-list'
import { useTouch } from '../hooks/use-touch'

interface TabViewItemCoords {
  vertical: boolean
  startX: number
  startY: number
  timeStamp: number
  position: number
  stop: boolean
}

export default defineComponent({
  name: 'fx-tab-view-item',
  props: {
    name: {
      type: String,
      required: true
    },
    subName: {
      type: String,
      default: null
    }
  },
  setup() {
    const root = ref<HTMLElement>()
    const vertical = inject('fxTabViewVertical', false)

    useListItem('swiper')
    useListItem('tabView')

    // onUpdated(() => {
    //   const $item = root.value as HTMLElement

    //   if ($item.offsetWidth === 0 || $item.offsetHeight === 0) {
    //     // 解决默认 hidden 的问题
    //     update()
    //   }
    // })

    let coords: TabViewItemCoords | null

    useTouch({
      el: root,
      onTouchStart(e) {
        const {
          scrollHeight,
          scrollTop,
          clientHeight,
          scrollLeft,
          scrollWidth,
          clientWidth
        } = e.currentTarget as HTMLElement

        const touch = e.touchObject

        if (
          (vertical &&
            (scrollHeight === scrollTop + clientHeight || scrollTop === 0)) ||
          (!vertical &&
            (scrollWidth === scrollLeft + clientWidth || scrollLeft === 0))
        ) {
          if (scrollHeight !== clientHeight || scrollWidth !== clientWidth) {
            coords = {
              vertical,
              position:
                (vertical && scrollTop === 0) || (!vertical && scrollLeft === 0)
                  ? 1
                  : 2,
              startX: touch.pageX,
              startY: touch.pageY,
              timeStamp: e.timeStamp,
              stop: false
            }
          }
        } else {
          // 滚动到中间，直接拒绝掉
          coords = {
            vertical,
            position: 1,
            startX: 0,
            startY: 0,
            timeStamp: e.timeStamp,
            stop: true
          }
          // e.stopPropagation()
        }
      },
      onTouchMove(e) {
        if (!coords) {
          return
        }

        if (coords.stop) {
          e.stopPropagation()
          return
        }

        const { pageX, pageY } = e.touchObject

        const offset = coords.vertical
          ? coords.startY - pageY
          : coords.startX - pageX

        if (
          (coords.position === 1 && offset > 0) ||
          (coords.position === 2 && offset < 0)
        ) {
          coords.stop = true
          e.stopPropagation()
        }
      },
      onTouchEnd() {
        coords = null
      }
    })

    return {
      root
    }
  }
})
</script>
