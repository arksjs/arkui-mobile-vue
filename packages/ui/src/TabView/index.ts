import { multiWithInstall } from '../helpers/with-install'
import TabView from './TabView.vue'
import TabViewItem from './TabViewItem.vue'

export { TabView, TabViewItem }
export default TabView
export const install = multiWithInstall(TabView, [TabViewItem])
