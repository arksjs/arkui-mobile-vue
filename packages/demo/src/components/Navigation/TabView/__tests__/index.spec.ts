import { mount, shallowMount } from '@vue/test-utils'
import { TabView, TabViewItem } from '@/TabView'

describe('TabView', () => {
  test('TabView should render default correctly', () => {
    const wrapper = shallowMount(TabView, {})

    expect(wrapper.html()).toMatchSnapshot()
  })

  test('TabViewItem should render default correctly', () => {
    const wrapper = mount(TabViewItem, {
      props: { name: 'item-1' },
      slots: {
        default: 'content'
      }
    })

    expect(wrapper.html()).toMatchSnapshot()
  })
})
