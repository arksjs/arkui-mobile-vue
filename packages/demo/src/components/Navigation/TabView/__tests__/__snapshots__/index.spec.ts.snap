// Vitest Snapshot v1

exports[`TabView > TabView should render default correctly 1`] = `
"<div class=\\"fx-tab-view\\">
  <div class=\\"fx-tab-view_header fx-horizontal-hairline\\">
    <tab-stub activevalue=\\"0\\" options=\\"\\" scrollthreshold=\\"4\\"></tab-stub>
  </div>
  <div class=\\"fx-tab-view_body\\">
    <swiper-stub indicatordots=\\"false\\" navigationbuttons=\\"false\\" autoplay=\\"false\\" interval=\\"5000\\" initialcircular=\\"false\\" activeindex=\\"0\\" initialvertical=\\"false\\" bounces=\\"false\\"></swiper-stub>
  </div>
</div>"
`;

exports[`TabView > TabViewItem should render default correctly 1`] = `"<div class=\\"fx-swiper-item fx-tab-view-item\\" data-name=\\"item-1\\">content</div>"`;
